lib/xdrgen/generators/go.rb

```diff
@@ -549,6 +531,259 @@ def render_encode_to_body(out, var, type, self_encode:)
+        struct.members.each do |m|
+          mn = name(m)
+          render_decode_from_body(out, "s.#{mn}", m.type, declared_variables: declared_variables, self_encode: false)
+        end
+        out.puts "  return n, nil"
+        out.puts "}"
+        out.break
+      end
+
+      def render_union_decode_from_interface(out, union)
+        name = name(union)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (u *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        render_decode_from_body(out, "u.#{name(union.discriminant)}", union.discriminant.type, declared_variables: [], self_encode: false)
+        switch_for(out, union, "u.#{name(union.discriminant)}") do |arm, kase|
+          out2 = StringIO.new
+          if arm.void?
+            out2.puts "// Void"
+          else
+            mn = name(arm)
+            type = arm.type
+            out2.puts "  u.#{mn} = new(#{reference arm.type})"
+            render_decode_from_body(out2, "(*u.#{mn})",type, declared_variables: [], self_encode: false)
+          end
+          out2.puts "  return n, nil"
+          out2.string
+        end
+        unless union.default_arm.present?
+          out.puts "  return n, fmt.Errorf(\"#{name(union.discriminant)} (#{reference union.discriminant.type}) switch value '%d' is not valid for union #{name}\", u.#{name(union.discriminant)})"
+        end
+        out.puts "}"
+        out.break
+      end
+
+      def render_enum_decode_from_interface(out, typedef)
+        name = name(typedef)
+        type = typedef
+        out.puts <<-EOS.strip_heredoc
+        // DecodeFrom decodes this value using the Decoder.
+        func (e *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {
+          v, n, err := d.DecodeInt()
+          if err != nil {
+            return n, err
+          }
+          if _, ok := #{private_name type}Map[v]; !ok {
+            return n, fmt.Errorf("'%d' is not a valid #{name} enum value", v)
+          }
+          *e = #{name}(v)
+          return n, nil
+        }
+        EOS
+      end
+
+      def render_typedef_decode_from_interface(out, typedef)
+        name = name(typedef)
+        type = typedef.declaration.type
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (s *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        var = "s"
+        sub_var_type = ""
+        case type
+        when AST::Typespecs::UnsignedHyper, AST::Typespecs::Hyper, AST::Typespecs::UnsignedInt, AST::Typespecs::Int, AST::Typespecs::String
+          sub_var_type = reference(type)
+        end
+        if (type.is_a?(AST::Typespecs::Opaque) && !type.fixed?) || (type.is_a?(AST::Typespecs::Simple) && type.sub_type == :var_array)
+            var = "(*s)"
+        end
+        if !sub_var_type.empty?
```

Comment:
nit: more idiomatic ruby way for `if !foo` is `unless foo`...
```suggestion
        unless sub_var_type.empty?
```